raft: improve group 0 reconfiguration failure handling

kbr-scylla opened the pull request:
Make it so that failures in `removenode`/`decommission` don't lead to reduced availability, and any leftovers in group 0 can be removed by `removenode`:
- In `removenode`, make the node a non-voter before removing it from the token ring. This removes the possibility of having a group 0 voting member which doesn't correspond to a token ring member. We can still be left with a non-voter, but that's doesn't reduce the availability of group 0.
- As above but for `decommission`.
- Make it possible to remove group 0 members that don't correspond to token ring members from group 0 using `removenode`. 
- Add an API to query the current group 0 configuration.

Fixes #11723.

kbr-scylla commented:
I'll add a commit for the docs tomorrow.

Request to reviewers: please review this promptly, it's important for Raft GA release in 5.2 which is planned to be branched next week.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3802/

kbr-scylla commented:
```
[2023-01-11T20:53:31.926Z] clang++ -MD -MT build/release/gen/rust/wasmtime_bindings.o -MF build/release/gen/rust/wasmtime_bindings.o.d -I/jenkins/workspace/releng/Scylla-CI/scylla/seastar/include -I/jenkins/workspace/releng/Scylla-CI/scylla/build/release/seastar/gen/include -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -march=westmere -Werror=unused-result -DSEASTAR_API_LEVEL=6 -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_HAVE_DPDK -DFMT_SHARED -DSEASTAR_BROKEN_SOURCE_LOCATION -I/usr/include/p11-kit-1 -Wl,--gc-sections -ffunction-sections -fdata-sections  -O3 -mllvm -inline-threshold=2500 -fno-slp-vectorize -DSCYLLA_BUILD_MODE=release -g -gz -iquote. -iquote build/release/gen --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/releng/Scylla-CI/scylla=.  -march=westmere -DBOOST_TEST_DYN_LINK   -DNOMINMAX -DNOMINMAX -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -Wno-c++11-narrowing -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lambda-capture -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-defaulted-function-deleted -Wno-redeclared-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-narrowing -Wno-array-bounds -Wno-nonnull -Wno-uninitialized -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DFMT_DEPRECATED_OSTREAM -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/release/gen/rust/wasmtime_bindings.o build/release/gen/rust/wasmtime_bindings.cc
[2023-01-11T20:53:31.926Z] build/release/gen/rust/wasmtime_bindings.cc:1:10: fatal error: 'rust/cxx.h' file not found
[2023-01-11T20:53:31.926Z] #include "rust/cxx.h"
[2023-01-11T20:53:31.926Z]          ^~~~~~~~~~~~
[2023-01-11T20:53:31.926Z] 1 error generated.
```
I think this was fixed on `master`, rebase should help

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -2344,19 +2344,54 @@ future<> storage_service::removenode(locator::host_id host_id, std::list<locator
             }
+
+            bool removed_from_token_ring = !endpoint_opt;
+        if (endpoint_opt) {
```
Forgot to fix indentation here, will add a commit after first round of review.

gleb-cloudius commented on test/topology_raft_disabled/test_raft_upgrade.py:
```diff
@@ -176,49 +176,6 @@ async def test_raft_upgrade_basic(manager: ManagerClient, random_tables: RandomT
     assert(table.full_name in rs[0].description)
 
 
-@pytest.mark.asyncio
```
I the document in the commit message public? Does it make sense to link to google docs at all?

kbr-scylla replied:
It's shared with ScyllaDB group. I think it doesn't hurt to add the link.

gleb-cloudius replied:
It does not, but imagine the annoyance of someone theoretical who read the commit message but cannot read the doc.

kostja replied:
I also happen to think it's better to put a link to apublic document. It's easy to add the text of the document to the issue or to a github wiki. It's approved now, and will likely stay frozen forever.

gleb-cloudius commented on service/raft/raft_group0.cc:
```diff
@@ -671,6 +671,30 @@ future<> raft_group0::finish_setup_after_join() {
     });
 }
 
+future<> raft_group0::become_nonvoter() {
```
Typo: "If the procedure fails before after step..."

gleb-cloudius replied:
Why `become_nonvoter()` does not call `make_nonvoter() ` with itself?

kbr-scylla replied:
The only practical difference between these functions is the logging events. `become_nonvoter` talks about "me" (this node), `make_nonvoter` talks about "them" (a different node). The common part which does the actual config change is already extracted to a separate private function, `make_raft_config_nonvoter`.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2015,13 +2018,27 @@ future<> storage_service::decommission() {
                 });
 
-                // Step 5: Start to sync data
+                // Step 4: Start to sync data
```
Looks like numbering steps is as bad idea as line numbers in Basic.

gleb-cloudius commented on api/api-doc/raft_group0.json:
```diff
@@ -0,0 +1,63 @@
+{
```
Better to implement https://github.com/scylladb/scylladb/issues/12008 instead adding new rest apis.

kbr-scylla replied:
#12008 is not a trivial project - the proposed table would contain all Raft groups and more information than just configuration. I'm also not familiar with the virtual table API that well. It would take me some time, which we don't have before the 5.2 release, to do it. The REST API doesn't hurt and it solves the current problem. I think we should avoid scope creep and keep this PR focused.

gleb-cloudius replied:
It may start as one group (that all we have anyways for now). I do not know much about virtual tables either, but we have a bunch so the infra should be in place to create more.

gleb-cloudius reviewed (commented):
Nitpicks mostly except API comment.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3812/

kbr-scylla commented:
Fixed typo in log message, causing formatting errors.
```
ERROR 2023-01-12 13:05:21,201 [shard 0] storage_service - service/storage_service.cc:2091 @operator(): failed to log message: fmt='decommission[{}}: left Raft group 0': fmt::v9::format_error (unmatched '}' in format string)
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3829/

kbr-scylla commented:
@asias @kostja @avikivity @tgrabiec ping, please review.

I'm working on a table for fetching raft config instead of the rest api, but the rest of the PR can be reviewed.

tgrabiec commented on service/raft/raft_group0.cc:
```diff
@@ -87,6 +87,10 @@ namespace service {
 static logging::logger upgrade_log("raft_group0_upgrade");
 
+// TODO: change the links from master to stable/5.2 after 5.2 is released
+static const auto raft_upgrade_doc = "https://docs.scylladb.com/master/architecture/raft.html#verifying-that-the-internal-raft-upgrade-procedure-finished-successfully";
```
Shouldn't we change it to 5.2 before 5.2 is released, so that the 5.2 binary refers to the "5.2" path? Should this depend on a variable set by the build system? It will become stale if it requires a manual process.

kbr-scylla replied:
You mean before we branch? I was planning to change it after we branch, on the 5.2 branch.
But maybe it makes sense to change it right now... Although currently the link works, if I change it to refer to non-existing (yet) path it will stop working.

Also, I don't actually know what the path will be :D right now we only have `stable` and `master`, where `stable` corresponds to 5.1, but once we release 5.2, I don't know what it will be. cc @annastuchlik 

Making it depend on a build system variable would be an interesting solution... Although we don't have the necessary scaffolding to do stuff like this right now.

tgrabiec replied:
We shouldn't change it to 5.2 before branching, of course. We should change it on 5.2 before releasing it. 

But I think we also should have an automated process which makes sure one you commit something to master and we branch, say 5.2, the code built from 5.2 will refer to the documentation for 5.2.

kostja replied:
https://docs.scylladb.com/stable/upgrade/ is a fairly stable link, but I wouldn't even use that. "Scylla release upgrade procedure" is easy to google and is more likely to stay relevant over time.

kbr-scylla replied:
Except "Scylla release upgrade procedure" is not what the link is pointing to, you could google "scylla internal raft upgrade", you can find https://docs.scylladb.com/stable/architecture/raft.html this way but it's not the first link and it's not pointing to the exact section that I wanted to point to here.

annastuchlik replied:
@kbr-scylla `stable `is and should always be the latest released stable version. As soon as we release 5.2, I'm going to change the configuration so that `stable` (default) is 5.2.

kbr-scylla replied:
We'll change this link to use `stable` after branching then.

kbr-scylla commented:
v2:
- rebased on `master`
- added a new commit to fix indentation in `removenode`
- removed the `api: add API to query Raft group 0 configuration` commit. Instead there is `db: system_keyspace: add a virtual table with raft configuration` (preceded with `db: system_keyspace: improve system.raft_snapshot_config schema` to make the schemas match)
- update the test to use the new table instead of the rest api.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3840/

kostja commented on service/raft/raft_group0.cc:
```diff
@@ -669,70 +669,8 @@ future<> raft_group0::finish_setup_after_join() {
-    if (!joined_group0()) {
-        on_internal_error(group0_log,
-            "leave_group0: we're fully upgraded to use Raft but didn't join group 0. Please report a bug.");
+    if (!(co_await raft_upgraded())) {
```
Sorry to bikeshed but I think English is playing a nasty trick here. The name is ambiguous - are we waiting for the non-raft cluster to upgrade to raft, or a raft cluster to upgrade to a better raft ? The name implies the second. I think raft_upgrade_complete() is less ambiguous, or upgrade_to_raft_complete(), or something along these lines.

kbr-scylla replied:
Renamed to `raft_upgrade_complete`.

kostja reviewed (commented):
https://docs.google.com/document/d/1Gvy2_ZezCa2XY5li05umOznIwjvIYFGmYpeZQg5RbZM/ - we could add this link to the github issue as a comment or as a wiki to avoid links to internal google doc. Alternatively, you could just copy the relevant bits into the commit comment and avoid the link altogether.

kostja commented on service/storage_service.cc:
```diff
@@ -2015,13 +2018,27 @@ future<> storage_service::decommission() {
+                    ss._group0->become_nonvoter().get();
+                    slogger.info("decommission[{}]: became a group 0 non-voter", uuid);
+                }
+
```
If we crash at this step, can decommission get reverted and we stay a non-voter forever?

asias replied:
If we crash here, the user can restart the node, the node will still be part of the ring. Will the raft code during restart, make the node a voter again?

kbr-scylla replied:
The node will become a voter again during restart, yes.

asias replied:
Good.

asias replied:
There is another tool we can use to revert the non-voter to voter in case of failure. 

There is a abort function attached to each node_ops_meta_data. The abort function will be called automatically if the decommission node does not update in a timrly manager (like a watchdog) to revert the decommission ops.

Perhaps we can call 'add the decommission node to a voter' in the abort function. This way, the decommission node will become voter automatically without user's operation. 

But on the other hand, without restarting the decommission node, it does not really matter if the decommission node is a voter or a non-voter, since we need to restart the decommission node to make it work properly again anyway. 

```
            auto meta = node_ops_meta_data(ops_uuid, coordinator, std::move(req.ignore_nodes), [this, coordinator, req = std::move(req)] () mutable {
                return mutate_token_metadata([this, coordinator, req = std::move(req)] (mutable_token_metadata_ptr tmptr) mutable {
                    for (auto& node : req.leaving_nodes) {
                        slogger.info("decommission[{}]: Removed node={} as leaving node, coordinator={}", req.ops_uuid, node, coordinator);
                        tmptr->del_leaving_endpoint(node);
                    }
                    return update_pending_ranges(tmptr, format("decommission {}", req.leaving_nodes)); 
                }); 
            },
            [this, ops_uuid] () mutable { node_ops_singal_abort(ops_uuid); });
            _node_ops.emplace(ops_uuid, std::move(meta));
```

kbr-scylla replied:
> But on the other hand, without restarting the decommission node, it does not really matter if the decommission node is a voter or a non-voter, since we need to restart the decommission node to make it work properly again anyway.


This.

And it can actually be considered a good thing that it is a non-voter while it is down.

kostja commented on service/raft/raft_group0.cc:
```diff
@@ -776,6 +776,10 @@ future<bool> raft_group0::wait_for_raft() {
     }
 
+    group0_log.info("Performing a group 0 read barrier...");
+    co_await _raft_gr.group0().read_barrier(&_abort_source);
```
I believe this was explicitly avoided before since we wanted the nodes to restart successfully in absence of a majority. I think we even have dtests for it. @gleb-cloudius

gleb-cloudius replied:
Yes. Good catch. I am not sure we have a test, but if we do not we should have one. It is fine to call the barrier during 'removenode' operation though before accessing the config.

kbr-scylla replied:
`wait_for_raft` is not called during restart.

kostja replied:
checked the code, this is the case indeed.

asias commented on service/storage_service.cc:
```diff
@@ -2368,7 +2383,20 @@ future<> storage_service::removenode(locator::host_id host_id, std::list<locator
+                ss._group0->make_nonvoter(raft_id).get();
+                slogger.info("removenode[{}]: made node {} a non-voter in group 0", uuid, raft_id);
+            }
+
```
In decommission, we make the node non-voter after sync data, here in removenode, we do this before sync data. Can we make they behave the same, i.e., making it non-voter after sync data? The step to sync data takes most of the time and has the best chance to fail.

kbr-scylla replied:
I did this on purpose, `removenode` assumes that the node is dead, so it's desirable to make the dead node a non-voter as soon as possible so we improve group 0's availability. There's no upside to keeping it as a voter for the duration of data sync. In `decommission` the situation is different, we're still alive, so it's desirable for us to keep being a voter until the last moment.

asias replied:
OK. Another point is that it is not a big deal even if we make the node leaving as non voter in case of failure.

asias commented on service/storage_service.cc:
```diff
@@ -2413,12 +2441,12 @@ future<> storage_service::removenode(locator::host_id host_id, std::list<locator
 
 
-                // Step 5: Announce the node has left
+                // Step 6: Announce the node has left
```
I.e., call ```ss._group0->make_nonvoter(raft_id).get()``` right before this step.

kbr-scylla replied:
See my answer above, I want `make_nonvoter` to happen early in `removenode`.

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -2647,10 +2648,120 @@ class clients_table : public streaming_virtual_table {
+            .set_comment("Currently operating RAFT configuration")
+            .with_version(system_keyspace::generate_schema_version(id))
+            .build();
+    }
```
As suggested in https://github.com/scylladb/scylladb/issues/12008 we will want to include more info (such as current term, leader etc) info this table. We may skip it for now but we have to have a plan how we will do it later. So what is the plan? Static columns?

kbr-scylla replied:
I was thinking static column too. Current term and leader need only a single row.
Also maybe `raft_config` is not an appropriate name if we want to include this info in the future (current term and leader are not part of the configuration if we want to use precise terminology). We could use something more generic like `raft_group_info`. Although using `raft_config` is not that a big problem.

gleb-cloudius replied:
For someone who does not know raft internals `raft_config` should not be confusing, though may be `raft_state` would be better.

kbr-scylla replied:
"state" is on the other hand too generic to me, arguably the raft log is also part of the state (for example), but it's kept in a separate table.

Meh, let's just keep it as `raft_config`.

gleb-cloudius replied:
Raft log *is* part of the state and one of the things I want to add to the table is its length in entries and bytes :)

Those other tables have different purpose. They are persisting the state of the raft needed for it to work. They are not for users to look at and may eventually be removed in favor of persisting directly in the commitlog.

I do not insist on the name though.

kbr-scylla replied:
Renamed to `raft_state` after all.

kbr-scylla commented:
v3:
- rename `raft_upgraded` to `raft_upgrade_complete`
- remove link to internal doc from commit message
- rename `raft_config` to `raft_state`, since we plan to extend this table in the future with more information. Write a comment at the definition of `raft_state_table`.

kbr-scylla commented:
> https://docs.google.com/document/d/1Gvy2_ZezCa2XY5li05umOznIwjvIYFGmYpeZQg5RbZM/ - we could add this link to the github issue as a comment or as a wiki to avoid links to internal google doc. Alternatively, you could just copy the relevant bits into the commit comment and avoid the link altogether.

Added the link to issue https://github.com/scylladb/scylladb/issues/11723, removed from commit message and cover letter.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3847/

kbr-scylla commented:
Failed to do a function rename fully, fixed.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3848/

tgrabiec commented:
Conflicts, please rebase.

kbr-scylla commented:
I rebased on current `next` because there's a conflict with Benny's patch, and I want CI to start running ASAP so we can queue this today

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3870/

tgrabiec commented on service/storage_service.cc:
```diff
@@ -2087,10 +2098,33 @@ future<> storage_service::decommission() {
+            // --- hence the `left_token_ring` check.
+            std::exception_ptr leave_group0_ex;
+            try {
+                if (raft_available && left_token_ring) {
```
Isn't left_token_ring always true here? If we didn't leave the ring, we will exit via throwing.
